NPE in io.quarkus:quarkus-maven-plugin:0.12.0:build

`quarkus-maven-plugin:build` fails with a `NullPointerException`:

````
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:0.12.0:build (default) on project kaiserkai-rest-impl: Failed to build a runnable JAR: Failed to build a runner jar: Failed to augment application classes: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.arc.deployment.ArcAnnotationProcessor#build threw an exception: java.lang.NullPointerException
````
The detailed stack trace is:

````
Caused by: java.lang.NullPointerException
    at io.quarkus.arc.processor.Beans.hasQualifier (Beans.java:405)
    at io.quarkus.arc.processor.BeanDeployment.findDisposer (BeanDeployment.java:694)
    at io.quarkus.arc.processor.BeanDeployment.findBeans (BeanDeployment.java:635)
    at io.quarkus.arc.processor.BeanDeployment.<init> (BeanDeployment.java:136)
    at io.quarkus.arc.processor.BeanProcessor.process (BeanProcessor.java:150)
    at io.quarkus.arc.deployment.ArcAnnotationProcessor.build (ArcAnnotationProcessor.java:237)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at io.quarkus.deployment.ExtensionLoader$1.execute (ExtensionLoader.java:506)
    at org.jboss.builder.BuildContext.run (BuildContext.java:413)
    at org.jboss.threads.ContextClassLoaderSavingRunnable.run (ContextClassLoaderSavingRunnable.java:35)
    at org.jboss.threads.EnhancedQueueExecutor.safeRun (EnhancedQueueExecutor.java:1998)
    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask (EnhancedQueueExecutor.java:1525)
    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run (EnhancedQueueExecutor.java:1416)
    at java.lang.Thread.run (Thread.java:748)
    at org.jboss.threads.JBossThread.run (JBossThread.java:479)

````

I don't fully understand what's going on here, but there seems to be a broken API contract: `io.quarkus.arc.processor.Beans.hasQualifier(ClassInfo, AnnotationInstance, Collection<AnnotationInstance>)` expects the third argument to be non-null, whereas `BeanDeployment.findDisposer()` calls it with a null value.

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Annotations.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Annotations.java
@@ -17,8 +17,12 @@
 package io.quarkus.arc.processor;
 
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
 import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationTarget.Kind;
 import org.jboss.jandex.DotName;
+import org.jboss.jandex.MethodInfo;
 
 public final class Annotations {
 
@@ -61,6 +65,17 @@ public final class Annotations {
             }
         }
         return false;
+    }
+
+    static Set<AnnotationInstance> getParameterAnnotations(BeanDeployment beanDeployment, MethodInfo method, int position) {
+        Set<AnnotationInstance> annotations = new HashSet<>();
+        for (AnnotationInstance annotation : beanDeployment.getAnnotations(method)) {
+            if (Kind.METHOD_PARAMETER == annotation.target().kind()
+                    && annotation.target().asMethodParameter().position() == position) {
+                annotations.add(annotation);
+            }
+        }
+        return annotations;
     }
 
 }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
@@ -691,15 +691,14 @@ public class BeanDeployment {
             if (disposer.getDeclaringBean().equals(declaringBean)) {
                 boolean hasQualifier = true;
                 for (AnnotationInstance qualifier : qualifiers) {
-                    if (!Beans.hasQualifier(getQualifier(qualifier.name()), qualifier, null)) {
+                    if (!Beans.hasQualifier(getQualifier(qualifier.name()), qualifier,
+                            disposer.getDisposedParameteterQualifiers())) {
                         hasQualifier = false;
                     }
                 }
-                if (hasQualifier && beanResolver.matches(beanType,
-                        disposer.getDisposerMethod().parameters().get(disposer.getDisposedParameter().position()))) {
+                if (hasQualifier && beanResolver.matches(beanType, disposer.getDiposedParameterType())) {
                     found.add(disposer);
                 }
-
             }
         }
         if (found.size() > 1) {

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
@@ -692,11 +692,11 @@ public class BeanDeployment {
                 boolean hasQualifier = true;
                 for (AnnotationInstance qualifier : qualifiers) {
                     if (!Beans.hasQualifier(getQualifier(qualifier.name()), qualifier,
-                            disposer.getDisposedParameteterQualifiers())) {
+                            disposer.getDisposedParameterQualifiers())) {
                         hasQualifier = false;
                     }
                 }
-                if (hasQualifier && beanResolver.matches(beanType, disposer.getDiposedParameterType())) {
+                if (hasQualifier && beanResolver.matches(beanType, disposer.getDisposedParameterType())) {
                     found.add(disposer);
                 }
             }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java
@@ -1501,7 +1501,7 @@ public class BeanGenerator extends AbstractGenerator {
             annotations = beanDeployment.getAnnotations(field);
         } else {
             MethodInfo method = injectionPoint.getTarget().asMethod();
-            annotations = InjectionPointInfo.getParameterAnnotations(beanDeployment,
+            annotations = Annotations.getParameterAnnotations(beanDeployment,
                     method, injectionPoint.getPosition());
         }
         for (AnnotationInstance annotation : annotations) {

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java
@@ -52,7 +52,7 @@ final class Beans {
     static BeanInfo createClassBean(ClassInfo beanClass, BeanDeployment beanDeployment) {
         Set<AnnotationInstance> qualifiers = new HashSet<>();
         ScopeInfo scope = null;
-        Set<Type> types = Types.getTypeClosure(beanClass, Collections.emptyMap(), beanDeployment);
+        Set<Type> types = Types.getClassBeanTypeClosure(beanClass, Collections.emptyMap(), beanDeployment);
         Integer alternativePriority = null;
         boolean isAlternative = false;
         List<StereotypeInfo> stereotypes = new ArrayList<>();
@@ -115,7 +115,7 @@ final class Beans {
             DisposerInfo disposer) {
         Set<AnnotationInstance> qualifiers = new HashSet<>();
         ScopeInfo scope = null;
-        Set<Type> types = Types.getTypeClosure(producerMethod, beanDeployment);
+        Set<Type> types = Types.getProducerMethodTypeClosure(producerMethod, beanDeployment);
         Integer alternativePriority = null;
         boolean isAlternative = false;
         List<StereotypeInfo> stereotypes = new ArrayList<>();
@@ -186,7 +186,7 @@ final class Beans {
             DisposerInfo disposer) {
         Set<AnnotationInstance> qualifiers = new HashSet<>();
         ScopeInfo scope = null;
-        Set<Type> types = Types.getTypeClosure(producerField, beanDeployment);
+        Set<Type> types = Types.getProducerFieldTypeClosure(producerField, beanDeployment);
         Integer alternativePriority = null;
         boolean isAlternative = false;
         List<StereotypeInfo> stereotypes = new ArrayList<>();

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DisposerInfo.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DisposerInfo.java
@@ -76,13 +76,13 @@ public class DisposerInfo {
         }
     }
 
-    Collection<AnnotationInstance> getDisposedParameteterQualifiers() {
+    Collection<AnnotationInstance> getDisposedParameterQualifiers() {
         return Annotations.getParameterAnnotations(declaringBean.getDeployment(), disposerMethod, disposedParameter.position())
                 .stream().filter(a -> declaringBean.getDeployment().getQualifier(a.name()) != null)
                 .collect(Collectors.toList());
     }
 
-    Type getDiposedParameterType() {
+    Type getDisposedParameterType() {
         return disposerMethod.parameters().get(disposedParameter.position());
     }
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DisposerInfo.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DisposerInfo.java
@@ -17,12 +17,15 @@
 package io.quarkus.arc.processor;
 
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.List;
+import java.util.stream.Collectors;
 import javax.enterprise.inject.spi.DefinitionException;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget.Kind;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.MethodParameterInfo;
+import org.jboss.jandex.Type;
 
 /**
  *
@@ -71,6 +74,16 @@ public class DisposerInfo {
         for (InjectionPointInfo injectionPoint : injection.injectionPoints) {
             Beans.resolveInjectionPoint(declaringBean.getDeployment(), null, injectionPoint, errors);
         }
+    }
+
+    Collection<AnnotationInstance> getDisposedParameteterQualifiers() {
+        return Annotations.getParameterAnnotations(declaringBean.getDeployment(), disposerMethod, disposedParameter.position())
+                .stream().filter(a -> declaringBean.getDeployment().getQualifier(a.name()) != null)
+                .collect(Collectors.toList());
+    }
+
+    Type getDiposedParameterType() {
+        return disposerMethod.parameters().get(disposedParameter.position());
     }
 
     MethodParameterInfo initDisposedParam(MethodInfo disposerMethod) {

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InjectionPointInfo.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InjectionPointInfo.java
@@ -26,7 +26,6 @@ import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget;
-import org.jboss.jandex.AnnotationTarget.Kind;
 import org.jboss.jandex.FieldInfo;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.Type;
@@ -62,7 +61,7 @@ public class InjectionPointInfo {
         for (ListIterator<Type> iterator = method.parameters().listIterator(); iterator.hasNext();) {
             Type paramType = iterator.next();
             int position = iterator.previousIndex();
-            Set<AnnotationInstance> paramAnnotations = getParameterAnnotations(beanDeployment, method, position);
+            Set<AnnotationInstance> paramAnnotations = Annotations.getParameterAnnotations(beanDeployment, method, position);
             if (skipPredicate != null && skipPredicate.test(paramAnnotations)) {
                 // Skip parameter, e.g. @Disposes
                 continue;
@@ -76,17 +75,6 @@ public class InjectionPointInfo {
             injectionPoints.add(new InjectionPointInfo(paramType, paramQualifiers, method, position));
         }
         return injectionPoints;
-    }
-
-    static Set<AnnotationInstance> getParameterAnnotations(BeanDeployment beanDeployment, MethodInfo method, int position) {
-        Set<AnnotationInstance> annotations = new HashSet<>();
-        for (AnnotationInstance annotation : beanDeployment.getAnnotations(method)) {
-            if (Kind.METHOD_PARAMETER.equals(annotation.target().kind())
-                    && annotation.target().asMethodParameter().position() == position) {
-                annotations.add(annotation);
-            }
-        }
-        return annotations;
     }
 
     private final TypeAndQualifiers typeAndQualifiers;

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
@@ -23,6 +23,7 @@ import io.quarkus.arc.WildcardTypeImpl;
 import io.quarkus.gizmo.BytecodeCreator;
 import io.quarkus.gizmo.MethodDescriptor;
 import io.quarkus.gizmo.ResultHandle;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -31,6 +32,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.FieldInfo;
@@ -143,56 +145,67 @@ final class Types {
         }
     }
 
-    static Set<Type> getTypeClosure(MethodInfo producerMethod, BeanDeployment beanDeployment) {
+    static Set<Type> getProducerMethodTypeClosure(MethodInfo producerMethod, BeanDeployment beanDeployment) {
+        Set<Type> types;
         Type returnType = producerMethod.returnType();
         if (returnType.kind() == Kind.PRIMITIVE) {
-            Set<Type> types = new HashSet<>();
+            types = new HashSet<>();
             types.add(returnType);
             types.add(OBJECT_TYPE);
             return types;
-        }
-        ClassInfo returnTypeClassInfo = beanDeployment.getIndex().getClassByName(returnType.name());
-        if (returnTypeClassInfo == null) {
-            throw new IllegalArgumentException(
-                    "Producer method return type not found in index: " + producerMethod.returnType().name());
-        }
-        if (Kind.CLASS.equals(returnType.kind())) {
-            return getTypeClosure(returnTypeClassInfo, Collections.emptyMap(), beanDeployment);
-        } else if (Kind.PARAMETERIZED_TYPE.equals(returnType.kind())) {
-            return getTypeClosure(returnTypeClassInfo,
-                    buildResolvedMap(returnType.asParameterizedType().arguments(), returnTypeClassInfo.typeParameters(),
-                            Collections.emptyMap()),
-                    beanDeployment);
         } else {
-            throw new IllegalArgumentException("Unsupported return type");
+            ClassInfo returnTypeClassInfo = beanDeployment.getIndex().getClassByName(returnType.name());
+            if (returnTypeClassInfo == null) {
+                throw new IllegalArgumentException(
+                        "Producer method return type not found in index: " + producerMethod.returnType().name());
+            }
+            if (Kind.CLASS.equals(returnType.kind())) {
+                types = getTypeClosure(returnTypeClassInfo, Collections.emptyMap(), beanDeployment);
+            } else if (Kind.PARAMETERIZED_TYPE.equals(returnType.kind())) {
+                types = getTypeClosure(returnTypeClassInfo,
+                        buildResolvedMap(returnType.asParameterizedType().arguments(), returnTypeClassInfo.typeParameters(),
+                                Collections.emptyMap()),
+                        beanDeployment);
+            } else {
+                throw new IllegalArgumentException("Unsupported return type");
+            }
         }
+        return restrictBeanTypes(types, beanDeployment.getAnnotations(producerMethod));
     }
 
-    static Set<Type> getTypeClosure(FieldInfo producerField, BeanDeployment beanDeployment) {
+    static Set<Type> getProducerFieldTypeClosure(FieldInfo producerField, BeanDeployment beanDeployment) {
+        Set<Type> types;
         Type fieldType = producerField.type();
         if (fieldType.kind() == Kind.PRIMITIVE) {
-            Set<Type> types = new HashSet<>();
+            types = new HashSet<>();
             types.add(fieldType);
             types.add(OBJECT_TYPE);
-            return types;
-        }
-        ClassInfo fieldClassInfo = beanDeployment.getIndex().getClassByName(producerField.type().name());
-        if (fieldClassInfo == null) {
-            throw new IllegalArgumentException("Producer field type not found in index: " + producerField.type().name());
-        }
-        if (Kind.CLASS.equals(fieldType.kind())) {
-            return getTypeClosure(fieldClassInfo, Collections.emptyMap(), beanDeployment);
-        } else if (Kind.PARAMETERIZED_TYPE.equals(fieldType.kind())) {
-            return getTypeClosure(fieldClassInfo,
-                    buildResolvedMap(fieldType.asParameterizedType().arguments(), fieldClassInfo.typeParameters(),
-                            Collections.emptyMap()),
-                    beanDeployment);
         } else {
-            throw new IllegalArgumentException("Unsupported return type");
+            ClassInfo fieldClassInfo = beanDeployment.getIndex().getClassByName(producerField.type().name());
+            if (fieldClassInfo == null) {
+                throw new IllegalArgumentException("Producer field type not found in index: " + producerField.type().name());
+            }
+            if (Kind.CLASS.equals(fieldType.kind())) {
+                types = getTypeClosure(fieldClassInfo, Collections.emptyMap(), beanDeployment);
+            } else if (Kind.PARAMETERIZED_TYPE.equals(fieldType.kind())) {
+                types = getTypeClosure(fieldClassInfo,
+                        buildResolvedMap(fieldType.asParameterizedType().arguments(), fieldClassInfo.typeParameters(),
+                                Collections.emptyMap()),
+                        beanDeployment);
+            } else {
+                throw new IllegalArgumentException("Unsupported return type");
+            }
         }
+        return restrictBeanTypes(types, beanDeployment.getAnnotations(producerField));
     }
 
-    static Set<Type> getTypeClosure(ClassInfo classInfo, Map<TypeVariable, Type> resolvedTypeParameters,
+    static Set<Type> getClassBeanTypeClosure(ClassInfo classInfo, Map<TypeVariable, Type> resolvedTypeParameters,
+            BeanDeployment beanDeployment) {
+        return restrictBeanTypes(getTypeClosure(classInfo, resolvedTypeParameters, beanDeployment),
+                beanDeployment.getAnnotations(classInfo));
+    }
+
+    private static Set<Type> getTypeClosure(ClassInfo classInfo, Map<TypeVariable, Type> resolvedTypeParameters,
             BeanDeployment beanDeployment) {
         Set<Type> types = new HashSet<>();
         List<TypeVariable> typeParameters = classInfo.typeParameters();
@@ -236,18 +249,26 @@ final class Types {
                 types.addAll(getTypeClosure(superClassInfo, resolved, beanDeployment));
             }
         }
+        return types;
+    }
 
-        // Bean types restriction
-        AnnotationInstance typed = classInfo.classAnnotations().stream().filter(a -> a.name().equals(DotNames.TYPED))
+    static Set<Type> restrictBeanTypes(Set<Type> types, Collection<AnnotationInstance> annotations) {
+        AnnotationInstance typed = annotations.stream().filter(a -> a.name().equals(DotNames.TYPED))
                 .findFirst().orElse(null);
         if (typed != null) {
-            Set<DotName> typedClasses = new HashSet<>();
-            for (Type type : typed.value().asClassArray()) {
-                typedClasses.add(type.name());
-            }
-            for (Iterator<Type> iterator = types.iterator(); iterator.hasNext();) {
-                if (!typedClasses.contains(iterator.next().name())) {
-                    iterator.remove();
+            AnnotationValue typedValue = typed.value();
+            if (typedValue == null) {
+                types.clear();
+                types.add(OBJECT_TYPE);
+            } else {
+                Set<DotName> typedClasses = new HashSet<>();
+                for (Type type : typedValue.asClassArray()) {
+                    typedClasses.add(type.name());
+                }
+                for (Iterator<Type> iterator = types.iterator(); iterator.hasNext();) {
+                    if (!typedClasses.contains(iterator.next().name())) {
+                        iterator.remove();
+                    }
                 }
             }
         }

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/producer/disposer/DisposerTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/producer/disposer/DisposerTest.java
@@ -56,7 +56,7 @@ public class DisposerTest {
                 }
             };
         }
-    }).around(new ArcTestContainer(StringProducer.class, LongProducer.class, BigDecimalProducer.class));
+    }).around(new ArcTestContainer(StringProducer.class, LongProducer.class, BigDecimalProducer.class, MyQualifier.class));
 
     @Test
     public void testDisposers() {

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/resolution/TypedTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/resolution/TypedTest.java
@@ -1,0 +1,87 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.quarkus.arc.test.resolution;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.ArcContainer;
+import io.quarkus.arc.InstanceHandle;
+import io.quarkus.arc.test.ArcTestContainer;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.Produces;
+import javax.enterprise.inject.Typed;
+import javax.inject.Singleton;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class TypedTest {
+
+    static final AtomicReference<String> EVENT = new AtomicReference<String>();
+
+    @Rule
+    public ArcTestContainer container = new ArcTestContainer(MyBean.class, MyOtherBean.class, Stage.class);
+
+    @Test
+    public void testEmptyTyped() throws IOException {
+        ArcContainer container = Arc.container();
+        assertFalse(container.instance(MyBean.class).isAvailable());
+        assertNull(EVENT.get());
+        container.beanManager().getEvent().fire("foo");
+        assertEquals("foo", EVENT.get());
+        InstanceHandle<Stage> stage = container.instance(Stage.class);
+        assertTrue(stage.isAvailable());
+        assertEquals("produced", stage.get().id);
+    }
+
+    @Typed // -> bean types = { Object.class }
+    @Singleton
+    static class MyBean {
+
+        void myObserver(@Observes String event) {
+            EVENT.set(event);
+        }
+
+    }
+
+    @Singleton
+    static class MyOtherBean {
+
+        @Produces
+        Stage myStage() {
+            return new Stage("produced");
+        }
+
+    }
+
+    @Typed
+    static class Stage {
+
+        final String id;
+
+        public Stage(String id) {
+            this.id = id;
+        }
+
+    }
+
+}

